### ISSUE TITLE
TestCaseSource with Test Method with params arguments

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#1883]
I thought I had seen this issue raised, but I cannot find it.  If this is a duplicate, I apologize.

If I have a test method:

Method (int arg1, string arg2, param object[] arg3)

I would hope to be able to past it an object from a TestCaseSource like:

new object[] {1, "abc", obj1, obj2, obj3, ...}

However, that results in a "Too many arguments provided" error.

Shouldn't TestCaseSourceAttribute detect the params and package all the remaining arguments as an object array rather then generating the error?

**Comment by ChrisMaddock** [comment]
Looks like it was just a PR without the issue. :smile: https://github.com/nunit/nunit/pull/1850

@cPetru - how's this going, do you manage to get things building in the end? Would be great to get this all untangled for the next release! 👍

**Comment by cPetru** [comment]
I had a small vacation. I will resume working on this issue, hopefully over
this weekend.

Pe 4 nov. 2016 10:44, "Chris Maddock" <notifications@github.com> a scris:

**Comment by ChrisMaddock** [comment]
@cPetru - Great, enjoy! The next release is a few months away, there's no rush. 😄

**Comment by CharliePoole** [comment]
@ChrisMaddock After reviewing an issue, you should close it (if rejecting) or add it to either Backlog or Discussion.

**Comment by ChrisMaddock** [comment]
@CharliePoole - Sorry, sometimes I use an app, which of course doesn't have Zenhub. Will try and remember to come back to things!

**Comment by CharliePoole** [comment]
@ChrisMaddock I use OctoDroid so I understand the problem. I wish they were Zenhub-aware!

**Comment by ChrisMaddock** [comment]
See https://github.com/nunit/nunit/issues/2268 for a more recent update on this issue.

**Comment by CharliePoole** [comment]
I was about to create an issue for this, but it's already here!

In my case I have a test method taking a single argument of type `params string[]`.

I can use `TestCaseAttribute` to call it, like this...

```c#
[TestCase("string1", "string2", "string3")]
```

But if I decide to use `TestCaseSourceAttribute, it won't work!

```c#
[TestCaseSource(nameof(MyData))]

...

TestCaseData[] MyData = new TestCaseData[] {
    new TestCaseData("string1", "string2", "string3")
};
```

The error says I can't have more than one argument.

This lack of consistency feels like a bug to me.

**Comment by stevenaw** [comment]
I have started preliminary investigation here and noticed that we have an analyzer rule which will require updating to accommodate this.

@mikkelbu @manfred-brands Is it better to release a rule relaxation before or after the framework side which makes it possible?

**Comment by manfred-brands** [comment]
@stevenaw The analyzer for `TestCaseAttribute` already supports `params`. It has a test for it:
```csharp
        [TestCase(1, 2, 3, 4)]
        public void Test(int a, int b, params int[] c) { }
```

The analyzer for `TesCaseSourceAttribute` it only checks there should be parameters as it cannot evaluate the code returning `TestCaseData` 

Summarizing, the analyzer should be ok.

**Comment by stevenaw** [comment]
Thanks @manfred-brands 

An analyzer did alert when I tried this yesterday for a test method with _only_ a params array and no preceding args. I can't recall the exact rule, but perhaps it was 2091. The issue it was alerting to was that the method required ">0" arguments. 

I'm happy to try to adjust that analyzer to not alert for that particular case, however I'm unclear if there's an order we'd prefer to tackle them in. I'm inclined to suggest the framework piece happen first but I'd appreciate hearing your thoughts.

**Comment by manfred-brands** [comment]
@stevenaw I agree that it is better to fix the framework first. The analyzer issue raised can be suppressed.
We had complains in the past for the other way around where the analyzer didn't allow something nunit does.
